lib/googleauth/credentials.rb

```diff
@@ -56,19 +59,24 @@ class Credentials
       def initialize(keyfile, options = {})
         scope = options[:scope]
         verify_keyfile_provided! keyfile
+        @project_id = options['project_id'] || options['project']
         if keyfile.is_a? Signet::OAuth2::Client
           @client = keyfile
+          @project_id ||= keyfile.instance_variable_get :@project_id
```

Comment:
What do you think of this?

```ruby
@project_id ||= keyfile.project_id if keyfile.respond_to? :project_id
```

Reply:
That's what I originally had before I switched to the present line - I thought it was simpler and more "tell. don't ask." What are your thoughts/reservations?

Reply:
If the implementation ever changes, and the variable is stored differently, then this can get out of sync. But, since the implementation is in the same library it is less likely to be a problem. We do this sometimes in google-cloud-* gems, but we try to keep it within the same class that uses the ivar and try to avoid it across classes.

Reply:
Generally instance variables are expected to be private to a class and subject to change. I'd try to avoid accessing them directly by name. Checking `respond_to?` before calling a method is pretty idiomatic, on the other hand, due to Ruby's duck typing nature.

Reply:
Thanks for the explanations!